diffmpm/material.py

```diff
@@ -133,6 +134,123 @@ def compute_stress(self, dstrain):
+        mu_ = material_properties["mu"]
+        critical_shear_rate = material_properties["critical_shear_rate"]
+        # Calculate the bulk modulus
+        bulk_modulus = youngs_modulus / (3.0 * (1.0 - 2.0 * poisson_ratio))
+        compressibility_multiplier_ = 1.0
+        # Special Material Properties
+        if "incompressible" in material_properties.keys():
+            incompressible = material_properties["incompressible"]
+            if incompressible:
+                compressibility_multiplier_ = 0.0
+        self.properties = {
+            **material_properties,
+            "bulk_modulus": bulk_modulus,
+            "compressibility_multiplier": compressibility_multiplier_,
+        }
+
+    def __repr__(self):
+        return f"Bingham(props={self.properties})"
+
+    # Initialise State Variables
+    def initialise_state_variables():
+        state_vars = {}
+        state_vars["pressure"] = 0.0
+        return state_vars
+
+    # State Variables
+    def state_variables():
+        return ["pressure"]
+
+    # Compute the pressure
+    def thermodynamic_pressure(self, volumetric_strain):
+        return -self.properties["bulk_modulus"] * volumetric_strain
+
+    # Compute the stress
+    def compute_stress(self, dstrain, particle, state_vars):
+        shear_rate_threshold = 1e-15
+        dirac_delta = jnp.array([0.0, 0.0, 0.0, 0.0, 0.0, 0.0]).reshape((6, 1))
+        if self.ndim == 1:
+            dirac_delta = dirac_delta.at[0, 0].set(1.0)
+        elif self.ndim == 2:
+            dirac_delta = dirac_delta.at[0:2, 0].set(1.0)
+
+        def compute_stress_per_particle(i):
+            strain_rate = particle.strain_rate[i]
+            strain_rate = strain_rate.at[-3:].set(strain_rate[-3:] * 0.5)
+            shear_rate_threshold = 1e-15
+            if self.properties["critical_shear_rate"] < shear_rate_threshold:
+                self.properties["critical_shear_rate"] = shear_rate_threshold
+            shear_rate = jnp.sqrt(
+                2.0
+                * (
+                    strain_rate.T @ (strain_rate)
+                    + strain_rate[-3:].T @ strain_rate[-3:]
+                )
+            )
+            apparent_viscosity = 0.0
+            if (shear_rate[0, 0] * shear_rate[0, 0]) > (
+                self.properties["critical_shear_rate"]
+                * self.properties["critical_shear_rate"]
+            ):
+                apparent_viscosity = 2.0 * (
+                    (self.properties["tau0"] / shear_rate[0, 0]) + self.properties["mu"]
+                )
+            tau = apparent_viscosity * strain_rate
+            trace_invariant2 = 0.5 * jnp.dot(tau[:3, 0], tau[:3, 0])
+            if trace_invariant2 < (self.properties["tau0"] * self.properties["tau0"]):
+                tau = tau.at[:].set(0)
+            state_vars["pressure"] += self.properties[
+                "compressibility_multiplier"
+            ] * self.thermodynamic_pressure(particle.dvolumetric_strain)
```

Comment:
Also, this is probably not going to work. `state_vars["pressure"]` is a `float` whereas `particle.dvolumetric_strain` is going to be an array of shape `(nparticles, 1)`